### ISSUE TITLE
Make Query/Builder more universal to allow data sources other than an sql db

### ISSUE THREAD

**Issue by JuhG** [laravel/ideas#1681]
Hey!

I'm trying to use Eloquent with data coming from a headless CMS by adding a new database connection. It's kind of working, but it feels hacky, since sql database is really baked into the core of the Database package. (For a good reason, of course.) Basically the whole Builder class boils down to the `toSql()` method, which returns a string, which is not suitable for any other type of data source.

How I imagine a future solution to this (possibly imaginary) problem, is that there is a missing step between the Builder and the Grammar class, that compiles the request to an array, which describes all the aspects of it, that can be turned into a string (in case of an sql db) or an api call (in case of a headless CMS) or anything else (in case of a markdown file or whatever).

Basically I think that the base Connection class or the Builder class shouldn't have opinion about my source, so PDO and SQL shouldn't be the backbone of these classes.

If I'm missing something obvious or this is already solved, then sorry, otherwise thanks for taking the time to read it.

**Comment by martinbean** [comment]
@JuhG The query builder is based on a PDO connection, which is why it only works against PDO-compatible database drivers like MySQL, Postgres, et al and not for databases like say, Mongo.

There’s nothing stopping you building a fluent query builder for your own data source, be that a RESTful API or non-relational database.
